### PR TITLE
REF: Fuse all the types

### PR DESCRIPTION
**Pull request by jbrockmendel**
Everything is passing locally, just want to run this through the CI for good measure before continuing down this path.

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -1153,7 +1153,7 @@ def check_int_bool(self, inplace):
                                                inplace=inplace, limit=limit,
                                                fill_value=fill_value,
                                                coerce=coerce,
-                                               downcast=downcast, mgr=mgr)
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py]: Edits here are unrelated, should be removed from this PR.

File: pandas/_libs/groupby_helper.pxi.in [hunk 2]
```diff
@@ -725,24 +725,37 @@ def group_min_{{name}}(ndarray[{{dest_type2}}, ndim=2] out,
...
-        {{dest_type2}} val, mval
-        ndarray[{{dest_type2}}, ndim=2] accum
+        groupby_t val, mval
+        ndarray[groupby_t, ndim=2] accum
         int64_t lab
 
     N, K = (<object> values).shape
     accum = np.empty_like(values)
-    accum.fill({{inf_val}})
+    if groupby_t is int64_t:
```
Comment (jreback) [inline pandas/_libs/groupby_helper.pxi.in]: can u make this more generic ? she what if we expand this to other int types?

**Comment by jbrockmendel** [comment]
Just pushed, moving some more functions over.  I think I should stop here for this PR before the diff gets out of hand.

Made a bug report to cython about a mysterious failure.

File: pandas/_libs/groupby_helper.pxi.in (hunk 2 above)
Comment (jreback) [reply pandas/_libs/groupby_helper.pxi.in]: this is a small cost of compile time (actually maybe nothing as cython is pretty smart). but i suppose can handle later.

Comment (jbrockmendel) [reply pandas/_libs/groupby_helper.pxi.in]: Fair enough.  Easy to implement if/when its actually needed.

**Comment by jbrockmendel** [comment]
Thoughts here?

File: pandas/_libs/algos_common_helper.pxi.in [hunk 3]
```diff
@@ -84,9 +80,9 @@ def diff_2d_{{name}}(ndarray[{{c_type}}, ndim=2] arr,
                     out[i, j] = arr[i, j] - arr[i, j - periods]
 
 
-def put2d_{{name}}_{{dest_type}}(ndarray[{{c_type}}, ndim=2, cast=True] values,
-                                 ndarray[int64_t] indexer, Py_ssize_t loc,
-                                 ndarray[{{dest_type2}}] out):
+def put2d_{{name}}_{{dest_type[:-2]}}(ndarray[{{c_type}}, ndim=2, cast=True] values,
+                                      ndarray[int64_t] indexer, Py_ssize_t loc,
```
Comment (jreback) [inline pandas/_libs/algos_common_helper.pxi.in]: this is a it obtuse can u make it more explicit (the slice)

File: pandas/_libs/groupby_helper.pxi.in [hunk 4]
```diff
@@ -638,7 +623,12 @@ def group_max_{{name}}(ndarray[{{dest_type2}}, ndim=2] out,
     nobs = np.zeros_like(out)
 
     maxx = np.empty_like(out)
-    maxx.fill(-{{inf_val}})
+    if groupby_t is int64_t:
```
Comment (jreback) [inline pandas/_libs/groupby_helper.pxi.in]: can u comment / add a Todo here

File: pandas/_libs/groupby_helper.pxi.in [hunk 5]
```diff
@@ -653,37 +643,44 @@ def group_max_{{name}}(ndarray[{{dest_type2}}, ndim=2] out,
...
-                    if val > maxx[lab, j]:
-                        maxx[lab, j] = val
+                if groupby_t is int64_t:
+                    if val != nan_val:
+                        nobs[lab, j] += 1
+                        if val > maxx[lab, j]:
+                            maxx[lab, j] = val
+                else:
+                    if val == val and val != nan_val:
+                        nobs[lab, j] += 1
```
Comment (jreback) [inline pandas/_libs/groupby_helper.pxi.in]: we should have a function that does the null comparisons though with the template code it is slightly 
tricky, maybe have a suite of isna_int, isna_float functions

File: pandas/_libs/groupby_helper.pxi.in [hunk 6]
```diff
@@ -653,37 +643,44 @@ def group_max_{{name}}(ndarray[{{dest_type2}}, ndim=2] out,
...
                 if nobs[i, j] == 0:
-                    out[i, j] = {{nan_val}}
+                    out[i, j] = nan_val
                 else:
                     out[i, j] = maxx[i, j]
 
 
+group_max_float64 = group_max["float64_t"]
+group_max_float32 = group_max["float32_t"]
+group_max_int64 = group_max["int64_t"]
```
Comment (jreback) [inline pandas/_libs/groupby_helper.pxi.in]: we DO need to expand these to all int/unit dtypes 
FYI

File: pandas/_libs/groupby_helper.pxi.in (hunk 2 above)
Comment (jreback) [reply pandas/_libs/groupby_helper.pxi.in]: see some comments above

File: pandas/_libs/sparse_op_helper.pxi.in [hunk 7]
```diff
@@ -106,10 +78,10 @@ def get_op(tup):
     ops_dict = {'add': '{0} + {1}',
                 'sub': '{0} - {1}',
                 'mul': '{0} * {1}',
-                'div': '__div_{2}({0}, {1})',
-                'mod': '__mod_{2}({0}, {1})',
-                'truediv': '__truediv_{2}({0}, {1})',
-                'floordiv': '__floordiv_{2}({0}, {1})',
+                'div': '__div({0}, {1})',
```
Comment (jreback) [inline pandas/_libs/sparse_op_helper.pxi.in]: these names are odd

maybe just call them __div__ and so on

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/groupby_helper.pxi.in (hunk 5 above)
Comment (jbrockmendel) [reply pandas/_libs/groupby_helper.pxi.in]: > we should have a function that does the null comparisons though with the template code 

I've been thinking about something similar, will give it a go in the next pass.

**Comment by jbrockmendel** [comment]
@datapythonista I don’t understand what is causing the travis failure.  Do you recognize it?

**Comment by datapythonista** [comment]
I'mc checking from my cell phone which makes it tricky to check, but if I see it correctly, this is the error: `pandas/_libs/algos_rank_helper.pxi.in:177:80: E501 line too long (80 > 79 characters)`.

File: pandas/_libs/groupby_helper.pxi.in (hunk 5 above)
Comment (jreback) [reply pandas/_libs/groupby_helper.pxi.in]: k

File: pandas/_libs/algos_rank_helper.pxi.in [hunk 8]
```diff
@@ -131,45 +131,20 @@ def rank_1d_{{dtype}}(object in_arr, ties_method='average',
     argsorted = _as.astype('i8')
 
     {{if dtype == 'object'}}
-    for i in range(n):
```
Comment (jreback) [inline pandas/_libs/algos_rank_helper.pxi.in]: is this all rolled into the generic rank? (except for the if around the nogil block)?

File: pandas/_libs/groupby_helper.pxi.in [hunk 9]
```diff
@@ -14,26 +14,22 @@ _int64_max = np.iinfo(np.int64).max
 
 {{py:
 
-# name, c_type, dest_type, dest_dtype
-dtypes = [('float64', 'float64_t', 'float64_t', 'np.float64'),
-          ('float32', 'float32_t', 'float32_t', 'np.float32')]
+# name, c_type
```
Comment (jreback) [inline pandas/_libs/groupby_helper.pxi.in]: is c_type the name we are calling this in other files? ideally try to be consistent

**Review by jreback — APPROVED** [review]
small question, other lgtm.

File: pandas/_libs/groupby_helper.pxi.in (hunk 9 above)
Comment (jbrockmendel) [reply pandas/_libs/groupby_helper.pxi.in]: There's some variation, but c_type is a common one.

File: pandas/_libs/algos_rank_helper.pxi.in (hunk 8 above)
Comment (jbrockmendel) [reply pandas/_libs/algos_rank_helper.pxi.in]: Right.  rank_2d already does it this way, so (besides being less verbose) this makes the patterns consistent.

**Comment by jreback** [comment]
thanks!

**Pull request merged**